// api-core/dist/src/edge/ApiEdgeQueryFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiEdgeQueryFilter = exports.ApiEdgeQueryFilterType = void 0;
var ApiEdgeQueryFilterType;
(function (ApiEdgeQueryFilterType) {
    ApiEdgeQueryFilterType[ApiEdgeQueryFilterType["Equals"] = 1] = "Equals";
    ApiEdgeQueryFilterType[ApiEdgeQueryFilterType["GreaterThanOrEquals"] = 2] = "GreaterThanOrEquals";
    ApiEdgeQueryFilterType[ApiEdgeQueryFilterType["GreaterThan"] = 3] = "GreaterThan";
    ApiEdgeQueryFilterType[ApiEdgeQueryFilterType["LowerThanOrEquals"] = 4] = "LowerThanOrEquals";
    ApiEdgeQueryFilterType[ApiEdgeQueryFilterType["LowerThan"] = 5] = "LowerThan";
    ApiEdgeQueryFilterType[ApiEdgeQueryFilterType["NotEquals"] = 6] = "NotEquals";
    ApiEdgeQueryFilterType[ApiEdgeQueryFilterType["Similar"] = 7] = "Similar";
    ApiEdgeQueryFilterType[ApiEdgeQueryFilterType["In"] = 8] = "In";
})(ApiEdgeQueryFilterType = exports.ApiEdgeQueryFilterType || (exports.ApiEdgeQueryFilterType = {}));
class ApiEdgeQueryFilter {
    constructor(field, type, value) {
        this.clone = () => new ApiEdgeQueryFilter(this.field, this.type, this.value);
        this.field = field;
        this.value = value;
        this.type = type;
    }
}
exports.ApiEdgeQueryFilter = ApiEdgeQueryFilter;
