// api-core/dist/src/edge/ApiEdgeAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiEdgeAction = exports.ApiEdgeActionTriggerKind = exports.ApiEdgeActionTrigger = void 0;
const ApiEdgeQueryType_1 = require("./ApiEdgeQueryType");
var ApiEdgeActionTrigger;
(function (ApiEdgeActionTrigger) {
    ApiEdgeActionTrigger[ApiEdgeActionTrigger["OutputQuery"] = 2] = "OutputQuery";
    ApiEdgeActionTrigger[ApiEdgeActionTrigger["SubQuery"] = 4] = "SubQuery";
    ApiEdgeActionTrigger[ApiEdgeActionTrigger["Method"] = 8] = "Method";
    ApiEdgeActionTrigger[ApiEdgeActionTrigger["Relation"] = 16] = "Relation";
    ApiEdgeActionTrigger[ApiEdgeActionTrigger["Query"] = 6] = "Query";
    ApiEdgeActionTrigger[ApiEdgeActionTrigger["Any"] = 30] = "Any";
})(ApiEdgeActionTrigger = exports.ApiEdgeActionTrigger || (exports.ApiEdgeActionTrigger = {}));
var ApiEdgeActionTriggerKind;
(function (ApiEdgeActionTriggerKind) {
    ApiEdgeActionTriggerKind[ApiEdgeActionTriggerKind["BeforeEvent"] = 0] = "BeforeEvent";
    ApiEdgeActionTriggerKind[ApiEdgeActionTriggerKind["AfterEvent"] = 1] = "AfterEvent";
})(ApiEdgeActionTriggerKind = exports.ApiEdgeActionTriggerKind || (exports.ApiEdgeActionTriggerKind = {}));
class ApiEdgeAction {
    constructor(name, execute, targetTypes = ApiEdgeQueryType_1.ApiEdgeQueryType.Any, triggerKind = ApiEdgeActionTriggerKind.BeforeEvent, triggers = ApiEdgeActionTrigger.Any, triggerNames = []) {
        this.triggerKind = ApiEdgeActionTriggerKind.BeforeEvent;
        this.targetTypes = ApiEdgeQueryType_1.ApiEdgeQueryType.Any;
        this.triggers = ApiEdgeActionTrigger.Any;
        this.triggerNames = [];
        this.inspect = () => {
            return `action{${this.name}}`;
        };
        this.name = name;
        this.triggers = triggers;
        this.execute = execute;
        this.targetTypes = targetTypes;
        this.triggerKind = triggerKind;
        this.triggerNames = triggerNames;
    }
}
exports.ApiEdgeAction = ApiEdgeAction;
